### PR TITLE
feat: batched and chunked insertion and deletion of relation tuples and UUID mappings

### PR DESCRIPTION
**Pull request by alnr**

### CODE REVIEW

**Review by aeneasr — CHANGES_REQUESTED** [review]
Generally looks good but I suggest we add tests that delete 500 tuples and create 500 tuples because my gut feeling says that we'll run into issues.

File: internal/persistence/sql/relationtuples.go [hunk 1]
```diff
@@ -165,25 +142,53 @@ func (p *Persister) whereQuery(ctx context.Context, q *pop.Query, rq *relationtu
...
+
+	args = make([]any, 0, 4*len(rs)+1)
+	ors := make([]string, 0, len(rs))
+	for _, rt := range rs {
+		switch s := rt.Subject.(type) {
+		case *relationtuple.SubjectID:
+			ors = append(ors, "(namespace = ? AND object = ? AND relation = ? AND subject_id = ? AND subject_set_namespace IS NULL AND subject_set_object IS NULL AND subject_set_relation IS NULL)")
+			args = append(args, rt.Namespace, rt.Object, rt.Relation, s.ID)
+		case *relationtuple.SubjectSet:
+			ors = append(ors, "(namespace = ? AND object = ? AND relation = ? AND subject_id IS NULL AND subject_set_namespace = ? AND subject_set_object = ? AND subject_set_relation = ?)")
```
[the remark this answers is not in the document]
Comment (alnr) [reply internal/persistence/sql/relationtuples.go:158]: Are you suggesting to use the pop query builder? Not possible because it can only `AND` several `WHERE` clauses together. There is no `OR`.

File: internal/persistence/sql/relationtuples.go [hunk 2]
```diff
@@ -165,25 +142,53 @@ func (p *Persister) whereQuery(ctx context.Context, q *pop.Query, rq *relationtu
...
+	}
+
 	return p.Transaction(ctx, func(ctx context.Context) error {
-		for _, r := range rs {
-			q := p.queryWithNetwork(ctx).
-				Where("namespace = ?", r.Namespace).
-				Where("object = ?", r.Object).
-				Where("relation = ?", r.Relation)
-			if err := p.whereSubject(ctx, q, r.Subject); err != nil {
+		for chunk := range slices.Chunk(rs, 500) {
```
[the remark this answers is not in the document]
Comment (alnr) [reply internal/persistence/sql/relationtuples.go:180]: 100 OK?

[the remark this answers is not in the document]
Comment (zepatrik) [reply internal/persistence/sql/relationtuples.go:180]: I think the more you can use in one query the better, because we'll save round-trips. Just add a test that reaches the maximum and see if it passes?

File: internal/persistence/sql/relationtuples.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (aeneasr) [reply internal/persistence/sql/relationtuples.go:158]: Sorry no, this was just a note to myself to comapre the queries!

File: internal/persistence/sql/relationtuples.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aeneasr) [reply internal/persistence/sql/relationtuples.go:180]: Yeah agree!

File: internal/persistence/sql/query_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+				Namespace: "ns3",
+				Object:    obj3,
+				Relation:  "rel3",
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	// parentheses are important here
+	assert.Equal(t, q, "DELETE FROM keto_relation_tuples WHERE ((namespace = ? AND object = ? AND relation = ? AND subject_id = ? AND subject_set_namespace IS NULL AND subject_set_object IS NULL AND subject_set_relation IS NULL) OR (namespace = ? AND object = ? AND relation = ? AND subject_id IS NULL AND subject_set_namespace = ? AND subject_set_object = ? AND subject_set_relation = ?)) AND nid = ?")
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply internal/persistence/sql/query_test.go:47]: I spoke to @zepatrik and he said there is no other way to do this, unfortunately. So I guess here too we just test what the maximum amount of parameters is and then set the limit accordingly?

File: internal/persistence/sql/relationtuples.go [hunk 4]
```diff
@@ -260,15 +265,63 @@ func (p *Persister) ExistsRelationTuples(ctx context.Context, query *relationtup
 	return exists, sqlcon.HandleError(err)
 }
 
+func buildInsert(commitTime time.Time, nid uuid.UUID, rs []*relationtuple.RelationTuple) (query string, args []any, err error) {
```
[the remark this answers is not in the document]
Comment (alnr) [reply internal/persistence/sql/relationtuples.go:268]: done

File: internal/persistence/sql/relationtuples.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alnr) [reply internal/persistence/sql/relationtuples.go]: Funnily, we don't seem to have an explicit limit on how many tuples we allow you to insert in one API call. gRPC has a max request size, but HTTP doesn't of course.

I've added a test which led me to discover that inserting the UUID mappings was batched (one statement) but not chunked. So I fixed that too.

Added comments and constants explaining chunk sizing.

IMO we should limit the max size also in the handlers. But the DB layer is capable of a lot!

**Review by jonas-jonas — APPROVED** [review]

File: internal/persistence/sql/relationtuples.go [hunk 5]
```diff
@@ -6,18 +6,29 @@ package sql
...
 	"github.com/ory/keto/internal/relationtuple"
 	"github.com/ory/keto/internal/x"
+	"github.com/ory/keto/ketoapi"
+)
+
+// Typical database limits for placeholders/bind vars are 1<<15 (32k, MySQL, SQLite) and 1<<16 (64k, PostgreSQL, CockroachDB).
+const (
+	chunkSizeInsertUUIDMappings = 15000 // two placeholders per mapping
+	chunkSizeInsertTuple        = 3000  // ten placeholders per tuple
+	chunkSizeDeleteTuple        = 100   // the database must build an expression tree for each chunk, so we must limit more agressively
```
Comment (zepatrik) [inline internal/persistence/sql/relationtuples.go:31]: What exactly do you mean by "build an expression tree" here?

File: internal/persistence/sql/relationtuples.go [hunk 6]
```diff
@@ -260,15 +272,63 @@ func (p *Persister) ExistsRelationTuples(ctx context.Context, query *relationtup
...
+	q.Grow(len(rs) * (len(placeholders) + len(separator)))
+	args = make([]any, 0, 10*len(rs))
+
+	for i, r := range rs {
+		if r.Subject == nil {
+			return "", nil, errors.WithStack(ketoapi.ErrNilSubject)
+		}
+
+		rt := &RelationTuple{
+			ID:         uuid.Must(uuid.NewV4()),
```
Comment (zepatrik) [inline internal/persistence/sql/relationtuples.go:293]: Use `gen_random_uuid()` on crdb?

File: internal/persistence/sql/relationtuples.go [hunk 7]
```diff
@@ -260,15 +272,63 @@ func (p *Persister) ExistsRelationTuples(ctx context.Context, query *relationtup
 	return exists, sqlcon.HandleError(err)
 }
 
+func buildInsert(commitTime time.Time, nid uuid.UUID, rs []*relationtuple.RelationTuple) (query string, args []any, err error) {
+	if len(rs) == 0 {
+		return "", nil, errors.WithStack(ketoapi.ErrMalformedInput)
+	}
+
+	var q strings.Builder
+	fmt.Fprintf(&q, "INSERT INTO %s (shard_id, nid, namespace, object, relation, subject_id, subject_set_namespace, subject_set_object, subject_set_relation, commit_time) VALUES ", (&RelationTuple{}).TableName())
```
Comment (zepatrik) [inline internal/persistence/sql/relationtuples.go:281]: As the nid and commit time are always the same, we could do this optimized query instead (simplified example):
```sql
INSERT INTO vals (nid, id, s)
SELECT 
    $1 AS nid, 
    gen_random_uuid() AS id, 
    v.s
FROM 
    (VALUES ($2), ($3), ($3)) AS v(s);
```
Not a blocker IMO, I'm happy either way.

File: internal/e2e/transaction_cases_test.go [hunk 8]
```diff
@@ -58,6 +60,53 @@ func runTransactionCases(c transactClient, m *namespaceTestManager) func(*testin
...
+			c.transactTuples(t, tuples, nil)
+
+			resp := c.queryTuple(t, &ketoapi.RelationQuery{
+				Namespace: &ns[0].Name,
+			})
+			// Use this instead, but currently unbearably huge output
+			// assert.ElementsMatch(t, tuples, resp.RelationTuples)
+			for i := range tuples {
+				require.Contains(t, resp.RelationTuples, tuples[i])
+			}
```
Comment (zepatrik) [inline internal/e2e/transaction_cases_test.go:100]: Sorry to break your test here, but it was not properly testing the functionality, as all RTs were the same.
It looks like you have to add pagination here to get the full list of relation tuples now (which you should have done from the beginning).

**Review by zepatrik — CHANGES_REQUESTED** [review]

File: internal/persistence/sql/relationtuples.go (hunk 5 above)
Comment (alnr) [reply internal/persistence/sql/relationtuples.go:31]: The DB parses the `DELETE WHERE` query and will construct a data structure for it. That's likely not an inexpensive operation especially if optimizations are performed on that structure afterwards. SQLite actually has a limit of 1000 nodes for this structure.

Inserts are a lot simpler in this regard, and we can have bigger queries.

File: internal/persistence/sql/relationtuples.go (hunk 6 above)
Comment (alnr) [reply internal/persistence/sql/relationtuples.go:293]: We could, but performance is not likely to increase a lot because these are global tables anyway.

File: internal/persistence/sql/relationtuples.go (hunk 7 above)
Comment (alnr) [reply internal/persistence/sql/relationtuples.go:281]: I don't think it is valid SQL to have `SELECT $1`. Either way, the chunk size is IMO large enough.

File: internal/e2e/transaction_cases_test.go (hunk 8 above)
Comment (alnr) [reply internal/e2e/transaction_cases_test.go]: pagination added!

**Review by zepatrik — APPROVED** [review]
Great 🎉

**Review by aeneasr — APPROVED** [review]

**Comment by alnr** [comment]
Alright after being led down some rabbit holes following red herrings (tired of animals yet?) I fixed the test and then marked it as slow.

But while investigating I also cleaned up some deprecated and arguably incorrect uses of the Go gRPC API and fixed a race condition. Didn't touch the non-test code.

Should be good to merge.

**Comment by AustinCase** [comment]
Hey! @aeneasr @alnr and @zepatrik saw this merged, but it looks like the last release thats available is v0.13-alpha from feb of last year. Are those being kept up to date?

**Comment by aeneasr** [comment]
We definitely owe a release, Patrik and me are both PTO but I‘ll put it on my list for when I‘m back end of Feb
